sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/HiveTypeCoercion.scala

```diff
@@ -59,6 +59,42 @@ object HiveTypeCoercion {
   }
 
   /**
+   * Implicit promote the DataType to StringType if
+   * one of the data type in (dt1, dt2) is StringType, and the other is not either
+   * BooleanType or BinaryType, the TightestCommonType should be StringType
+   * eg: 1. CaseWhenLike  case when ... then dt1 else dt2 end
+   *     2. Coalesce(null, dt1, dt2)
+
+   */
+  val promoteToStringTypeOfTwo: (DataType, DataType) => Option[DataType] = {
+    case (t1: StringType, t2: DataType) if (t2 != BinaryType && t2 != BooleanType) =>
```

Comment:
yes, @cloud-fan
you can refer the chart at the bottom of page in https://cwiki.apache.org/confluence/display/Hive/LanguageManual+Types 

This is only for primitive type.